spark/src/main/java/org/apache/iceberg/spark/source/Reader.java

```diff
@@ -533,69 +552,27 @@ private static UnsafeProjection projection(Schema finalSchema, Schema readSchema
       return CloseableIterable.transform(
           asSparkRows, APPLY_PROJECTION.bind(projection(readSchema, tableSchema))::invoke);
     }
-  }
 
-  private static class PartitionRowConverter implements Function<StructLike, InternalRow> {
-    private final DataType[] types;
-    private final int[] positions;
-    private final Class<?>[] javaTypes;
-    private final GenericInternalRow reusedRow;
-
-    PartitionRowConverter(Schema partitionSchema, PartitionSpec spec) {
-      StructType partitionType = SparkSchemaUtil.convert(partitionSchema);
-      StructField[] fields = partitionType.fields();
-
-      this.types = new DataType[fields.length];
-      this.positions = new int[types.length];
-      this.javaTypes = new Class<?>[types.length];
-      this.reusedRow = new GenericInternalRow(types.length);
-
-      List<PartitionField> partitionFields = spec.fields();
-      for (int rowIndex = 0; rowIndex < fields.length; rowIndex += 1) {
-        this.types[rowIndex] = fields[rowIndex].dataType();
-
-        int sourceId = partitionSchema.columns().get(rowIndex).fieldId();
-        for (int specIndex = 0; specIndex < partitionFields.size(); specIndex += 1) {
-          PartitionField field = spec.fields().get(specIndex);
-          if (field.sourceId() == sourceId && "identity".equals(field.transform().toString())) {
-            positions[rowIndex] = specIndex;
-            javaTypes[rowIndex] = spec.javaClasses()[specIndex];
-            break;
-          }
-        }
+    private DataType getPartitionType(Schema partitionSchema, String fieldName) {
+      if (partitionSchema.findField(fieldName) != null) {
+        return SparkSchemaUtil.convert(partitionSchema.findField(fieldName).type());
+      } else {
+        return SparkSchemaUtil.convert(tableSchema.findField(fieldName).type());
       }
     }
 
-    @Override
-    public InternalRow apply(StructLike tuple) {
-      for (int i = 0; i < types.length; i += 1) {
-        Object value = tuple.get(positions[i], javaTypes[i]);
-        if (value != null) {
-          reusedRow.update(i, convert(value, types[i]));
-        } else {
-          reusedRow.setNullAt(i);
+    private Object getPartitionValue(StructLike schema, int index, DataType sparkType, Class<?> javaType) {
```

Comment:
Can you restore the original `convert` method that was removed and just add support for null values to it? That would reduce the amount of churn in this PR.

Reply:
Adding back the `convert` method implies adding back the full `PartitionRowConverter` class. Are you ok with this?

Reply:
No, that's a static method. We can just move it out of the PartitionRowConverter class.

Reply:
Ok. Got it. :+1: